feat: add a BillingConnector to send funds to a Billing contract on L2

pcarranzav opened the pull request.

abarmat commented on contracts/Billing.sol:
```diff
@@ -59,9 +66,11 @@ contract Billing is IBilling, Governed {
     constructor(
         address _gateway,
```
This is unrelated to the scope of adding the token gateway (bridge) but I'd use the opportunity to rename `gateway` to `agent` or something different to avoid mistakes, the role is really a funds collection agent. In addition I think I'd make it so we can configure multiple agents.

pcarranzav replied:
Good idea. How about `collector`? (and using a mapping of authorized collectors)

abarmat commented on contracts/Billing.sol:
```diff
@@ -1,18 +1,17 @@
 // SPDX-License-Identifier: MIT
 
-pragma solidity ^0.8.4;
+pragma solidity ^0.8.16;
```
Nice to use the latest solidity version

abarmat commented on contracts/Billing.sol:
```diff
@@ -110,6 +121,17 @@ contract Billing is IBilling, Governed {
+     * @param _amount  Amount of tokens to add
+     */
+    function addFromL1(address _user, uint256 _amount) external override onlyL2TokenGateway {
+        userBalances[_user] = userBalances[_user] + _amount;
```
We could maybe leave this in `_add()`:
```
        userBalances[_user] = userBalances[_user] + _amount;
        emit TokensAdded(_user, _amount);
```
and have a `pullAndAdd()` with:
```
        require(_amount != 0, "Must add more than 0");
        require(_user != address(0), "user != 0");
        require(graphToken.transferFrom(_from, address(this), _amount), "Add transfer failed");
        _add(_user, _amount);
```

abarmat commented on contracts/BillingConnector.sol:
```diff
@@ -0,0 +1,155 @@
+     * @param _token  Token address of the token that was accidentally sent to the contract
+     * @param _amount  Amount of tokens to pull
+     */
+    function rescueTokens(
```
Maybe we can abstract this into a Rescatable class

abarmat commented on contracts/BillingConnector.sol:
```diff
@@ -0,0 +1,155 @@
+     * @dev Sets the L2 Billing address
+     * @param _l2Billing New address for the L2 Billing contract
+     */
+    function setL2Billing(address _l2Billing) external override onlyGovernor {
```
This is good as long as `onlyGovernor` is always the Council to avoid anyone to circumvent the gateway allowlist

abarmat commented on contracts/BillingConnector.sol:
```diff
@@ -0,0 +1,155 @@
+    ) external payable override {
+        require(_amount != 0, "Must add more than 0");
+        require(_to != address(0), "destination != 0");
+        require(graphToken.transferFrom(msg.sender, address(this), _amount), "Add transfer failed");
```
This might be a good opportunity to use the Permit feature from the GRT. We could add a `addToL2WithSig` so we can get the user to do everything with no token approval transaction, sign message, then send final transaction.

abarmat reviewed (commented):
This is looking very good! I left a few comments

abarmat commented:
@IainM22 @nenadjaja @brandonleekramer you can see a draft **BillingConnector** to bridge funds from L1 to an L2 billing

pcarranzav commented:
Thanks for the comments @abarmat, I've implemented your suggestions and fixed the existing tests. I still need to add new tests for the new contract and features.

tmigone commented on contracts/Billing.sol:
```diff
@@ -41,54 +44,59 @@ contract Billing is IBilling, Governed {
+        _setCollector(_collector, _enabled);
+    }
+
+    function setL2TokenGateway(address _l2TokenGateway) external onlyGovernor {
```
```suggestion
    function setL2TokenGateway(address _l2TokenGateway) external override onlyGovernor {
```

tmigone commented on contracts/IBilling.sol:
```diff
@@ -1,13 +1,14 @@
      */
-    function setGateway(address _newGateway) external; // onlyGateway or onlyGovernor, or something
+    function setCollector(address _collector, bool _enabled) external; // onlyGovernor
 
```
```suggestion

    /**
     * @dev Sets the L2 token gateway address
     * @param _l2TokenGateway New address for the L2 token gateway
     */
    function setL2TokenGateway(address _l2TokenGateway) external;

```

tmigone commented on contracts/Billing.sol:
```diff
@@ -41,54 +44,59 @@ contract Billing is IBilling, Governed {
+    function setCollector(address _collector, bool _enabled) external override onlyGovernor {
+        _setCollector(_collector, _enabled);
+    }
+
```
```suggestion
    
    /**
     * @dev Sets the L2 token gateway address
     * @param _l2TokenGateway New address for the L2 token gateway
     */
```

tmigone commented on contracts/BillingConnector.sol:
```diff
@@ -0,0 +1,195 @@
+    /**
+     * @dev Tokens sent to the Billing contract on L2
+     */
+    event TokensSentToL2(address _from, address _to, uint256 _amount);
```
```suggestion
    event TokensSentToL2(address indexed _from, address indexed _to, uint256 _amount);
```

Maybe we want to index the addresses?

tmigone reviewed (commented):
🚀 left a few minor suggestions

tmigone commented on contracts/BillingConnector.sol:
```diff
@@ -0,0 +1,265 @@
+    /**
+     * @dev Request sent to the Billing contract on L2 to remove tokens from the balance
+     */
+    event TokensRemovedOnL2(address indexed _from, address indexed _to, uint256 _amount);
```
I wonder if we should rename this event to something else given that as described below there is no guarantee that the tx on L2 will succeed. Maybe `RequestTokensRemovedOnL2` or something similar

pcarranzav replied:
Good point, replaced for `RemovalRequestSentToL2`

tmigone reviewed (approved):
👍
